### ISSUE TITLE
Incorrect execution of a complex query

### ISSUE THREAD

**Issue by Lexey** [linq2db/linq2db#413]
The following query:

``` cs
            var instrumentCodes = db.GetTable<T1>()
                .Where(_ => _.InstrumentCode.StartsWith(cond ? "aaa" : "bbb") && _.CreateDate <= uptoDate)
                .Join(db.GetTable<T2>(), _ => _.InstrumentId, _ => _.InstrumentId, (ins, idx) => idx.IndexId)
                .Join(db.GetTable<T3>(), _ => _, _ => _.IndexId, (idx, w) => w.InstrumentId)
                .Join(db.GetTable<T1>(), _ => _, _ => _.InstrumentId
                    , (w, ins) => ins.SourceInstrumentCode)
                .Where(_ => _ != null)
                .Distinct()
                .OrderBy(_ => _)
                .ToList();
```

results in a list with the null item inside (although it should be filtered out).
The generated SQL is:

``` sql
DEBUG 2016-08-17 12:50:24.0203 DataConnection: -- Server SqlServer.2008
DECLARE @p1 NVarChar(5) -- String
SET     @p1 = N'XXXX%'
DECLARE @uptoDate DateTime2
SET     @uptoDate = '2016-08-17'

SELECT DISTINCT
    [ins].[SourceInstrumentCode],
    [ins].[SourceInstrumentCode] as [SourceInstrumentCode1]
FROM
    [dbo].[T1] [ins1]
        INNER JOIN [dbo].[T2] [idx] ON [ins1].[Instrument_Id] = [idx].[Instrument_Id]
        INNER JOIN [dbo].[T3] [w] ON [idx].[Index_Id] = [w].[Index_Id]
        INNER JOIN [dbo].[T1] [ins] ON [w].[Instrument_Id] = [ins].[Instrument_Id]
WHERE
    [ins1].[InstrumentCode] LIKE @p1 ESCAPE '~' AND [ins1].[CreateDate] <= @uptoDate
ORDER BY
    [ins].[SourceInstrumentCode]
```

The **[ins].[SourceInstrumentCode] as [SourceInstrumentCode1]** also looks questionable.

Replacing

``` cs
                .Join(db.GetTable<T1>(), _ => _, _ => _.InstrumentId
                    , (w, ins) => ins.SourceInstrumentCode)
                .Where(_ => _ != null)
```

with

``` cs
                .Join(db.GetTable<T1>().Where(_ => _.SourceInstrumentCode != null), _ => _, _ => _.InstrumentId
                    , (w, ins) => ins.SourceInstrumentCode)
```

workarounds the issue.

**Comment by sdanyliv** [comment]
@Lexey, it is fundamentally to write method chains instead of linq expressions? I can not read or even understand this without doping ;)

**Comment by ili** [comment]
Yes... we do have a problem here :(
this is the same as #416

**Comment by MaceWindu** [comment]
Looks like already fixed #1944 

generated sql:

```sql
SELECT DISTINCT
    [ins].[SourceInstrumentCode]
FROM
    [T1] [_]
        INNER JOIN [T2] [idx] ON [_].[InstrumentId] = [idx].[InstrumentId]
        INNER JOIN [T3] [w] ON [idx].[IndexId] = [w].[IndexId]
        INNER JOIN [T1] [ins] ON [w].[InstrumentId] = [ins].[InstrumentId]
WHERE
    [ins].[SourceInstrumentCode] IS NOT NULL AND [_].[InstrumentCode] LIKE @cond ESCAPE N'~' AND
    [_].[CreateDate] <= @uptoDate
ORDER BY
    [ins].[SourceInstrumentCode]
```
